### DISCUSSION TITLE
Cut silence (below specified level)

### DISCUSSION THREAD

**Discussion by iSerganov** [savonet/liquidsoap#4059] [Q&A] [answered]
Hey community!

Can you please help me to find a way to cut silence in the beginning and in the end of a track?
E.g. if loudness level falls below `-45dB` cut the beginning (or the end of a track). 

I found [ffmpeg.filter.silenceremove](https://www.liquidsoap.info/doc-dev/reference-extras.html#ffmpeg.filter.silenceremove) filter in Liquidsoap documentation. However I am not really sure how can I use it...

Thanks in advance!

**Comment by vitoyucepi** [comment]
Hi @iSerganov,
You're probably looking for the `blank.*` function group:
- [`blank.detect`](https://www.liquidsoap.info/doc-2.2.5/reference.html#blank.detect)
- [`blank.eat`](https://www.liquidsoap.info/doc-2.2.5/reference.html#blank.eat)
- [`blank.skip`](https://www.liquidsoap.info/doc-2.2.5/reference.html#blank.skip)
- [`blank.strip`](https://www.liquidsoap.info/doc-2.2.5/reference.html#blank.strip)

You may also want to check out the examples in [the book](https://www.liquidsoap.info/book/book.pdf).

I should also mention that autocue may provide a better experience.

**Comment by iSerganov** [comment]
Thanks for info provided, @vitoyucepi!
It looks like **autocue** is exactly what I need... Could you please point me to a doc or example where I can configure and enable it for cutting silence, (and maybe smart crossfading) purposes?

**Comment by gAlleb** [comment, marked as answer]
Hi @iSerganov. From this issue https://github.com/savonet/liquidsoap/issues/4057 I've got an impression that you have familiarized yourself with `autocue`

`settings.autocue.cue_file.silence := -45.0`
About this setting: LU below integrated track loudness for cue-in & cue-out points (silence removal at beginning & end of a track) (default: -42.0)

The above setting is for `autocue` by MoonBase59.

If we are talking about internal `autocue` then we have this settings:
`settings.autocue.internal.cue_in_threshold` - defaults to  -34.0
`settings.autocue.internal.cue_out_threshold` - defaults to  -42.0

You can see all settings for internal autocue here -> https://github.com/savonet/liquidsoap/blob/27dda43947cdb39fca16a9fb435dc9f45bee1e4b/src/libs/autocue.liq and for `autocue` by Moonbase59 here -> https://github.com/Moonbase59/autocue

<hr>

1. Internal autocue will calculate all the fade_* values for you and its setup as easy as:

```ruby

enable_autocue_metadata()
s = playlist("music.m3u8")
s = amplify(1.,override="liq_amplify", s)
s = crossfade(s)
output(s)

```

2. Autocue.cue_file won't calculate fade_* values, defaults are: `fade_in = 0.1`, `fade_out = 2.5`. Which will default to a more segue-like "crossfading". E.g. : 

```
# Fading/crossing/segueing
def custom_crossfade(old, new) =

        add(normalize=false, [
          fade.in(
            initial_metadata=new.metadata,
            duration=settings.autocue.cue_file.fade_in(),
            new.source
          ),
          fade.out(
            initial_metadata=old.metadata,
            duration=settings.autocue.cue_file.fade_out(),
            old.source
          )
        ])

end

radio = cross(
  duration=settings.autocue.cue_file.fade_out(),
  custom_crossfade,
  radio
)

```

**Comment by iSerganov** [comment]
That's absolutely makes sense!

And the option No. 2 is exactly what I was looking for! I need to make `fade_in`/`fade_out` values configurable rather than automatically computed. 

As to the `settings.autocue.cue_file.silence := -45.0`, I was looking at it, but it wasn't clear where exactly it detects and cuts silence (end, start or middle of the file), so thanks for clarifying it.

That's a great and super helpful explanation which absolutely answers my question.

Thank you @gAlleb!
